design/design.md

```diff
@@ -0,0 +1,380 @@
+* `process()`, which is called after a request has been received, passes to the `JobQueue`
+a coroutine that calls the appropriate handler. `process()` also creates
+another `CallData` object to service any additional requests, since the current
+object is busy processing a request. Note, `doProcess()` itself
+does not execute the handler, but simply posts a coroutine to the `JobQueue` and
+returns.
+* `finish()`, is called after the response has been successfully sent, and
+  performs any necessary cleanup or post processing. In most cases, no post
+processing is necessary, so the base class provides a default implementation
+that does nothing. After `finish()` is executed, the `CallData` object destroys
+itself.
+
+In the setup of the event loop, a `CallData` object for each rpc method is
+created, which causes gRPC to listen for each type of request.
+When an event occurs, `CompletionQueue::Next()` returns a pointer to a
+`CallData` object as a tag, and the event loop calls `Proceed()` on the returned object. 
+If the event was reception of a request, `Proceed()` will call `process()`,
+which places a coroutine on the `JobQueue`, creates a new `CallData` object for
+this request type, and returns immediately (the coroutine executes the handler, which
+populates and sends the response, some time later); once the coroutine is
+executed by the `JobQueue` and the response has been sent, a pointer to this `CallData` object will be placed back on
+the `CompletionQueue`. When `CompletionQueue::Next()` returns a pointer to a `CallData`
+object for which a response has been sent, `Proceed()` will call `finish()` and
+then delete the object. Note that each tag sent to the `CompletionQueue` (through
+RPC operations in `makeListener()` or `process()`) will be delivered out of the `CompletionQueue` by a
+call to `CompletionQueue::Next()`, regardless of whether the operation
+succeeded or not. Success here means that this operation completed in the normal
+valid manner. This frees us from having to store `CallData` objects in a
+container, since `CompletionQueue` is a container of pointers to all `CallData` objects,
+and every pointer will eventually be returned from `CompletionQueue::Next()`,
+meaning every `CallData` object will be deleted eventually.
+
+```
+  void EventLoop() {
+    // Spawn new CallData instances to serve new clients.
+    new AccountInfoCallData(&service_, completion_queue_.get(), app_);
+    new FeeCallData(&service_, completion_queue_.get(), app_);
+    ///... Additional CallData instances for additional rpc methods
+
+    void* tag;  // uniquely identifies a request.
+    bool ok;
+    while (true) {
+      // Block waiting to read the next event from the completion queue. The
+      // event is uniquely identified by its tag, which is the memory address
+      // of a CallData instance.
+      completion_queue_->Next(&tag, &ok);
+      static_cast<CallData*>(tag)->Proceed();
```

Comment:
What do you think of something like this?:
```
    auto cd = static_cast<CallData*>(tag);
    if (cd->complete())
    {
        delete cd;
    }
    else
    {
        cd->clone(); // adds to completion queue
        cd->process();
    }
```
I like that the lifetime management is easier to see (although `clone` isn't great, I'd love to explicitly add to the Completion queue, but that may be hard). It gets rid of the `delete this;` and hidden object creation in `process`, it also gets rid of the state machine and `Proceed`.

Reply:
Changed the event loop so that lifetime management is easier to see, and removed the `Proceed` function